src/hotspot/share/opto/addnode.cpp

```diff
@@ -1095,185 +1096,154 @@ Node* MaxNode::build_min_max_diff_with_zero(Node* a, Node* b, bool is_max, const
+// If n is an integer addition of x with a non-TOP constant C, return <x, C>.
+// Otherwise, return <nullptr, 0>.
+static ConstAddOperands as_add_with_constant(Node* n) {
+  ConstAddOperands null(nullptr, 0);
+  if (n->Opcode() != Op_AddI) {
+    return null;
+  }
+  Node* x = n->in(1);
+  Node* c = n->in(2);
+  if (!c->is_Con()) {
+    return null;
+  }
+  const Type* c_type = c->bottom_type();
+  if (c_type == Type::TOP) {
+    return null;
+  }
+  return ConstAddOperands(x, c_type->is_int()->get_con());
+}
 
-    // Transform MAX2(x + c0, MAX2(x + c1, z)) into MAX2(x + MAX2(c0, c1), z)
-    // if x == y and the additions can't overflow.
-    if (x == y && tx != nullptr &&
-        !can_overflow(tx, x_off) &&
-        !can_overflow(tx, y_off)) {
-      return new MaxINode(phase->transform(new AddINode(x, phase->intcon(MAX2(x_off, y_off)))), r->in(2));
+Node* MaxNode::IdealI(PhaseGVN* phase, bool can_reshape) {
+  int opcode = Opcode();
+  assert(opcode == Op_MinI || opcode == Op_MaxI, "Unexpected opcode");
+  // Defer handling of op(x, x) to constant/identity propagation.
+  if (in(1) == in(2)) {
+    return nullptr;
+  }
+  // Try to transform the following pattern, in any of its four possible
+  // permutations induced by op's commutativity:
+  //     op(op(add(inner, inner_off), inner_other), add(outer, outer_off))
+  // into
+  //     op(add(inner, op(inner_off, outer_off)), inner_other),
+  // where:
+  //     op is either MinI or MaxI, and
+  //     inner == outer, and
+  //     the additions cannot overflow.
+  for (uint inner_op_index = 1; inner_op_index <= 2; inner_op_index++) {
+    if (in(inner_op_index)->Opcode() != opcode) {
+      continue;
     }
-  } else {
-    // Transform MAX2(x + c0, y + c1) into x + MAX2(c0, c1)
-    // if x == y and the additions can't overflow.
-    if (x == y && tx != nullptr &&
-        !can_overflow(tx, x_off) &&
-        !can_overflow(tx, y_off)) {
-      return new AddINode(x, phase->intcon(MAX2(x_off, y_off)));
+    Node* outer_add = in(inner_op_index == 1 ? 2 : 1);
+    ConstAddOperands outer_add_operands = as_add_with_constant(outer_add);
+    if (outer_add_operands.first == nullptr) {
+      continue;
+    }
+    Node* outer = outer_add_operands.first;
+    jint outer_off = outer_add_operands.second;
+    // One operand is a MinI/MaxI and the other is an integer addition with
+    // constant. Test the operands of the inner MinI/MaxI.
+    for (uint inner_add_index = 1; inner_add_index <= 2; inner_add_index++) {
+      Node* inner_op = in(inner_op_index);
+      Node* inner_add = inner_op->in(inner_add_index);
+      ConstAddOperands inner_add_operands = as_add_with_constant(inner_add);
+      if (inner_add_operands.first == nullptr) {
+        continue;
+      }
+      Node* inner = inner_add_operands.first;
+      jint inner_off = inner_add_operands.second;
+      // Try to extract the inner add.
+      Node* add_extracted = extract_add(phase, inner, inner_off, outer, outer_off);
```

Comment:
Optional: You could also leave `outer_add_operands` and `inner_add_operands` packed in the `Pair`, and pass it as such into `extract_add`. Could reduce the number of lines a bit here.